File: typescript/timestep/db_connection.ts
/** Database connection management for Timestep. */

import { getPgliteDir } from './app_dir.ts';

export enum DatabaseType {
  POSTGRESQL = 'postgresql',
  PGLITE = 'pglite',
  NONE = 'none',
}

export interface DatabaseConnectionOptions {
  connectionString?: string;
  usePglite?: boolean;
  pglitePath?: string;
}

export class DatabaseConnection {
  private connection: any = null;
  private dbType: DatabaseType = DatabaseType.NONE;
  private connectionString?: string;
  private usePglite: boolean;
  private pglitePath: string;

  constructor(options: DatabaseConnectionOptions = {}) {
    if (options.connectionString) {
      // PostgreSQL mode
      this.connectionString = options.connectionString;
      this.usePglite = false;
      this.pglitePath = '';
    } else if (options.usePglite) {
      // PGLite mode - path is required
      if (!options.pglitePath) {
        throw new Error('pglitePath is required when usePglite=true');
      }
      this.connectionString = undefined;
      this.usePglite = true;
      this.pglitePath = options.pglitePath;
    } else {
      // No configuration provided
      throw new Error(
        'Either connectionString must be provided for PostgreSQL, ' +
        'or usePglite=true with pglitePath for PGLite'
      );
    }
  }

  private _getDefaultPglitePath(): string {
    // Use app_dir module to match Python's get_app_dir() implementation
    return getPgliteDir();
  }

  async connect(): Promise<boolean> {
    /**
     * Connect to database.
     */
    if (this.connectionString && !this.usePglite) {
      try {
        return await this.connectPostgreSQL();
      } catch (e) {
        throw new Error(`Failed to connect to PostgreSQL: ${e}`);
      }
    }

    if (this.usePglite) {
      try {
        return await this.connectPglite();
      } catch (e) {
        throw new Error(`Failed to connect to PGLite: ${e}`);
      }
    }

    throw new Error('No connection configuration provided');
  }

  private async connectPostgreSQL(): Promise<boolean> {
    try {
      // Dynamic import to avoid requiring pg at module load time
      const pg = await import('pg');
      const { Pool } = pg;

      // Parse connection string and create pool
      this.connection = new Pool({
        connectionString: this.connectionString,
      });

      this.dbType = DatabaseType.POSTGRESQL;

      // Test connection
      await this.connection.query('SELECT 1');
      
      // Initialize schema
      const { initializeSchema } = await import('./schema.ts');
      await initializeSchema(this.connection);
      
      return true;
    } catch (e: any) {
      if (e.code === 'MODULE_NOT_FOUND' || e.message?.includes('Cannot find module')) {
        throw new Error(
          'pg is required for PostgreSQL support. Install it with: npm install pg @types/pg'
        );
      }
      throw new Error(`Failed to connect to PostgreSQL: ${e.message}`);
    }
  }

  private async connectPglite(): Promise<boolean> {
    try {
      // Dynamic import to avoid requiring @electric-sql/pglite at module load time
      const { PGlite } = await import('@electric-sql/pglite');
      const fs = await import('fs/promises');
      const path = await import('path');

      // Ensure directory exists
      const pgliteDir = path.dirname(this.pglitePath);
      try {
        await fs.mkdir(pgliteDir, { recursive: true });
      } catch (e: any) {
        // Ignore if directory already exists
        if (e.code !== 'EEXIST') {
          throw e;
        }
      }

      // Use 'idb' mode for better concurrency support
      // This allows multiple connections to the same database
      this.connection = new PGlite(this.pglitePath, { dataDir: this.pglitePath });
      this.dbType = DatabaseType.PGLITE;

      // Initialize database (creates if doesn't exist)
      await this.connection.waitReady;

      // Initialize schema (with error handling for concurrent access)
      try {
        const { initializeSchema } = await import('./schema.ts');
        await initializeSchema(this.connection);
      } catch (schemaError: any) {
        // If schema initialization fails due to concurrent access, that's okay
        // The table might already exist from another connection
        if (!schemaError.message?.includes('already exists') && 
            !schemaError.message?.includes('duplicate') &&
            !schemaError.message?.includes('Aborted')) {
          throw schemaError;
        }
      }

      // Test connection
      await this.connection.query('SELECT 1');
      return true;
    } catch (e: any) {
      if (e.code === 'MODULE_NOT_FOUND' || e.message?.includes('Cannot find module')) {
        throw new Error(
          '@electric-sql/pglite is required for PGLite support. Install it with: npm install @electric-sql/pglite'
        );
      }
      // Provide more helpful error message for WASM abort
      if (e.message?.includes('Aborted')) {
        throw new Error(
          `Failed to connect to PGLite: Database may be locked or corrupted. ` +
          `Try using a unique database path per test or ensure proper cleanup. ` +
          `Original error: ${e.message}`
        );
      }
      throw new Error(`Failed to connect to PGLite: ${e.message}`);
    }
  }

  async disconnect(): Promise<void> {
    if (this.connection) {
      if (this.dbType === DatabaseType.POSTGRESQL) {
        await this.connection.end();
      } else if (this.dbType === DatabaseType.PGLITE) {
        await this.connection.close();
      }
      this.connection = null;
      this.dbType = DatabaseType.NONE;
    }
  }

  get isConnected(): boolean {
    return this.connection !== null && this.dbType !== DatabaseType.NONE;
  }

  get connectionHandle(): any {
    if (!this.isConnected) {
      throw new Error('Database not connected. Call connect() first.');
    }
    return this.connection;
  }

  get type(): DatabaseType {
    return this.dbType;
  }

  async query(text: string, params?: any[]): Promise<any> {
    if (this.dbType === DatabaseType.POSTGRESQL) {
      return await this.connection.query(text, params);
    } else if (this.dbType === DatabaseType.PGLITE) {
      return await this.connection.query(text, params);
    } else {
      throw new Error(`Query not implemented for ${this.dbType}`);
    }
  }
}

File: python/timestep/db_connection.py
"""Database connection management for Timestep."""

import os
from pathlib import Path
from typing import Optional, Any
from enum import Enum
from .app_dir import get_pglite_dir
from .pglite_sidecar.client import PGliteSidecarClient


class DatabaseType(Enum):
    """Database type enumeration."""
    POSTGRESQL = "postgresql"
    PGLITE = "pglite"
    NONE = "none"


class DatabaseConnection:
    """Manages database connections for Timestep."""
    
    def __init__(
        self,
        connection_string: Optional[str] = None,
        use_pglite: Optional[bool] = None,
        pglite_path: Optional[str] = None
    ):
        """
        Initialize database connection.
        
        Args:
            connection_string: PostgreSQL connection string (e.g., postgresql://user:pass@host/db)
            use_pglite: Whether to use PGLite (required if connection_string is None)
            pglite_path: Path for PGLite data directory (required if use_pglite is True)
        """
        if connection_string:
            # PostgreSQL mode
            self.connection_string = connection_string
            self.use_pglite = False
            self.pglite_path = None
        elif use_pglite:
            # PGLite mode - path is required
            if not pglite_path:
                raise ValueError("pglite_path is required when use_pglite=True")
            self.connection_string = None
            self.use_pglite = True
            self.pglite_path = pglite_path
        else:
            # No configuration provided
            raise ValueError(
                "Either connection_string must be provided for PostgreSQL, "
                "or use_pglite=True with pglite_path for PGLite"
            )
        self._connection: Optional[Any] = None
        self._db_type: DatabaseType = DatabaseType.NONE
        self._pglite_sidecar: Optional[PGliteSidecarClient] = None
        self._pglite_started = False
        
    async def connect(self) -> bool:
        """
        Connect to database. Returns True if successful, False otherwise.
        
        Returns:
            True if connection successful, False otherwise
        """
        if self.connection_string and not self.use_pglite:
            try:
                return await self._connect_postgresql()
            except Exception as e:
                raise ConnectionError(f"Failed to connect to PostgreSQL: {e}")
        
        if self.use_pglite:
            try:
                return await self._connect_pglite()
            except Exception as e:
                raise ConnectionError(f"Failed to connect to PGLite: {e}")
        
        raise ValueError("No connection configuration provided")
    
    async def _connect_postgresql(self) -> bool:
        """Connect to PostgreSQL database."""
        try:
            import asyncpg
            
            # Parse connection string
            self._connection = await asyncpg.connect(self.connection_string)
            self._db_type = DatabaseType.POSTGRESQL
            
            # Test connection
            await self._connection.fetchval("SELECT 1")
            
            # Initialize schema (pass self so it uses our unified interface)
            from .schema import initialize_schema
            await initialize_schema(self)
            
            return True
        except ImportError:
            raise ImportError(
                "asyncpg is required for PostgreSQL support. "
                "Install it with: pip install asyncpg"
            )
        except Exception as e:
            raise ConnectionError(f"Failed to connect to PostgreSQL: {e}")
    
    async def _connect_pglite(self) -> bool:
        """Connect to PGLite database via sidecar."""
        try:
            # Initialize PGLite path
            pglite_path = Path(self.pglite_path)
            pglite_path.mkdir(parents=True, exist_ok=True)
            
            # Start sidecar
            self._pglite_sidecar = PGliteSidecarClient(str(pglite_path))
            await self._pglite_sidecar.start()
            self._pglite_started = True
            
            # Store path for reference, connection is the sidecar
            self._connection = self._pglite_sidecar
            self._db_type = DatabaseType.PGLITE
            
            # Test connection
            await self._pglite_sidecar.query("SELECT 1", None)
            
            # Initialize schema (pass self so it uses our unified interface)
            from .schema import initialize_schema
            await initialize_schema(self)
            
            return True
        except Exception as e:
            raise ConnectionError(f"Failed to connect to PGLite: {e}")
    
    async def disconnect(self) -> None:
        """Disconnect from database."""
        if self._connection:
            if self._db_type == DatabaseType.POSTGRESQL:
                await self._connection.close()
            elif self._db_type == DatabaseType.PGLITE:
                # Stop the sidecar
                if self._pglite_sidecar:
                    await self._pglite_sidecar.stop()
                    self._pglite_sidecar = None
                    self._pglite_started = False
            self._connection = None
            self._db_type = DatabaseType.NONE
    
    @property
    def is_connected(self) -> bool:
        """Check if database is connected."""
        return self._connection is not None and self._db_type != DatabaseType.NONE
    
    @property
    def connection(self) -> Any:
        """Get database connection."""
        if not self.is_connected:
            raise RuntimeError("Database not connected. Call connect() first.")
        return self._connection
    
    @property
    def db_type(self) -> DatabaseType:
        """Get database type."""
        return self._db_type
    
    async def execute(self, query: str, *args) -> Any:
        """Execute a query."""
        if self._db_type == DatabaseType.POSTGRESQL:
            return await self._connection.execute(query, *args)
        elif self._db_type == DatabaseType.PGLITE:
            if not self._pglite_sidecar:
                raise RuntimeError("PGLite sidecar not initialized")
            # PGLite uses query method which returns result
            params = list(args) if args else None
            result = await self._pglite_sidecar.query(query, params)
            # Return rowCount if available, otherwise None
            return result.get('rowCount', None) if isinstance(result, dict) else None
        else:
            raise NotImplementedError(f"Execute not implemented for {self._db_type}")
    
    async def fetch(self, query: str, *args) -> list:
        """Fetch rows from a query."""
        if self._db_type == DatabaseType.POSTGRESQL:
            return await self._connection.fetch(query, *args)
        elif self._db_type == DatabaseType.PGLITE:
            if not self._pglite_sidecar:
                raise RuntimeError("PGLite sidecar not initialized")
            params = list(args) if args else None
            return await self._pglite_sidecar.fetch(query, params)
        else:
            raise NotImplementedError(f"Fetch not implemented for {self._db_type}")
    
    async def fetchrow(self, query: str, *args) -> Optional[dict]:
        """Fetch a single row from a query."""
        if self._db_type == DatabaseType.POSTGRESQL:
            return await self._connection.fetchrow(query, *args)
        elif self._db_type == DatabaseType.PGLITE:
            if not self._pglite_sidecar:
                raise RuntimeError("PGLite sidecar not initialized")
            params = list(args) if args else None
            return await self._pglite_sidecar.fetchrow(query, params)
        else:
            raise NotImplementedError(f"Fetchrow not implemented for {self._db_type}")
    
    async def fetchval(self, query: str, *args) -> Any:
        """Fetch a single value from a query."""
        if self._db_type == DatabaseType.POSTGRESQL:
            return await self._connection.fetchval(query, *args)
        elif self._db_type == DatabaseType.PGLITE:
            if not self._pglite_sidecar:
                raise RuntimeError("PGLite sidecar not initialized")
            params = list(args) if args else None
            return await self._pglite_sidecar.fetchval(query, params)
        else:
            raise NotImplementedError(f"Fetchval not implemented for {self._db_type}")

